packages/mesh/src/Mesh.ts

```diff
@@ -82,7 +82,7 @@ export class Mesh extends Container
          * Can be shared between multiple Mesh objects.
          * @member {PIXI.Shader|PIXI.MeshMaterial}
          */
-        this.shader = shader;
+        this.shader = shader as MeshMaterial;
```

Comment:
Okay, what's the point then :). It's better to change the type of `this.shader` and act accordingly.

Reply:
If change the type of `this.shader` to `Shader|MeshMaterial` we will need to cast in a few places then, because `Shader` doesn't have properties like `texture`, `alpha`, 'tint' etc.

Reply:
Well, that's what we need to do. If `Shader` is a supported mesh shader, then we can't blindly assume it'll work?

Reply:
I think we can assume that it will work because for `Shader` or `MeshMaterial` with `program` we call `_renderDefault` that not require `texture`, `tint` etc.

Reply:
Yeah, so wherever you require the shader to be `MeshMaterial` you need to cast it there.

Reply:
There's always `Mesh<Shader>` and `Mesh<MeshMaterial>` to prevent casting.

Reply:
I don't understand what you mean @SukantPal. Can you elaborate

Reply:
```ts
class Mesh<T extends Shader = MeshMaterial> {
  material: T
  render() {
    if (this.material instanceof MeshMaterial) {} // implicit cast into MeshMaterial
    else {}
  }
}
```

You'd still have to check if the material is `MeshMaterial` or not.

Reply:
`instanceof` is not a cast, it's a runtime type check. I think that's fine.

Reply:
Yeah, I think Generics is the way to go for this. What do you think @SerG-Y?

Reply:
I am not sure that it will work in this specific case, just an example we have getters and setters for `tint` and `texture` in `Mesh` class and with generic we will need cast it to `MeshMaterial` in all places. 
Just take a look at the following piece of code from `Mesh`. When we pass `Shader` and set `texture` it will not break anything because of JS but in TS we need cast it to `MeshMaterial` and in the code, we have few places where we should cast it to `MeshMaterial` when we will use Generics. Hope my thoughts are clear for you) What do you think guys?
```js
get texture(): Texture
{
    return this.shader.texture;
}

set texture(value: Texture)
{
    this.shader.texture = value;
}
```

Reply:
```ts
/** Null for non-MeshMaterial shaders. */
get texture(): Texture
{
  return 'texture' in this.shader ? this.shader.texture : null;
}
```

Reply:
These issues you're describing are mostly internal. What's more important here is that externally we handle this Mesh class better. I think generics solves that problem, but yeah we'll need to resolve the internal bits with a combo of casting and/or runtime checking. That's okay, IMO.

Reply:
Are you agree with the following type of casting?
```ts
const shader = this.shader as unknown as MeshMaterial;
```
it's not possible directly cast to `MeshMaterial` because of the next error. If you know a better way how to do it, please let me know.
![image](https://user-images.githubusercontent.com/12532719/107614158-6efb1980-6c52-11eb-82a0-8d1c6af55035.png)

Reply:
Sure, not ideal but that's fine.